File: src/ios/swift/SettingsView.swift
import SwiftUI

struct SettingsView: View {

    // MARK: - Loading

    @AppStorage("ios.lazyRooms")
    private var lazyRooms = true

    @AppStorage("ios.lazyTextures")
    private var lazyTextures = true

    @AppStorage("ios.lazyAudio")
    private var lazyAudio = true

    @AppStorage("ios.loadType")
    private var loadType = 0

    // MARK: - Display

    @AppStorage("ios.touchControls")
    private var touchControls = true

    @AppStorage("ios.showFPS")
    private var showFPS = true

    @AppStorage("ios.widescreen")
    private var widescreen = false

    @AppStorage("ios.widescreenAspect")
    private var widescreenAspect = 1.7777778

    @AppStorage("ios.windowWidth")
    private var windowWidth = 0

    @AppStorage("ios.windowHeight")
    private var windowHeight = 0

    @AppStorage("ios.renderer")
    private var renderer = 0

    @AppStorage("ios.headless")
    private var headless = false

    // MARK: - Performance

    @AppStorage("ios.speed")
    private var speed = 1.0

    @AppStorage("ios.fastForwardSpeed")
    private var fastForwardSpeed = 0.0

    @AppStorage("ios.profilerFramesBetween")
    private var profilerFramesBetween = 0

    @AppStorage("ios.opcodeProfiler")
    private var opcodeProfiler = false

    @AppStorage("ios.metal")
    private var metal = true;

    // MARK: - Debugging

    @AppStorage("ios.debug")
    private var debug = false

    @AppStorage("ios.disassemble")
    private var disassemble = false

    @AppStorage("ios.alwaysLogUnknownFunctions")
    private var alwaysLogUnknownFunctions = false

    @AppStorage("ios.alwaysLogStubbedFunctions")
    private var alwaysLogStubbedFunctions = false

    @AppStorage("ios.disableLogColours")
    private var disableLogColours = false

    @AppStorage("ios.traceFrames")
    private var traceFrames = false

    @AppStorage("ios.traceEventInherited")
    private var traceEventInherited = false

    // MARK: - Printing

    @AppStorage("ios.printRooms")
    private var printRooms = false

    @AppStorage("ios.printObjects")
    private var printObjects = false

    @AppStorage("ios.printShaders")
    private var printShaders = false

    @AppStorage("ios.printDeclaredFunctions")
    private var printDeclaredFunctions = false

    @AppStorage("ios.printUnknownFunctions")
    private var printUnknownFunctions = false

    // MARK: - Frame Control

    @AppStorage("ios.exitAtFrame")
    private var exitAtFrame = -1

    @AppStorage("ios.traceBytecodeAfterFrame")
    private var traceBytecodeAfterFrame = 0

    // MARK: - Random Seed

    @AppStorage("ios.hasSeed")
    private var hasSeed = false

    @AppStorage("ios.seed")
    private var seed = 0

    // MARK: - Input Recording

    @AppStorage("ios.recordInputsPath")
    private var recordInputsPath = ""

    @AppStorage("ios.playbackInputsPath")
    private var playbackInputsPath = ""

    // MARK: - Misc

    @AppStorage("ios.eagerRooms")
    private var eagerRooms = ""

    @AppStorage("ios.dumpJsonFilePattern")
    private var dumpJsonFilePattern = ""

    var body: some View {

        NavigationStack {

            Form {

                // MARK: Loading

                Section("Loading") {

                    Toggle(
                        "Lazy room loading",
                        isOn: $lazyRooms
                    )

                    Toggle(
                        "Lazy texture loading",
                        isOn: $lazyTextures
                    )

                    Toggle(
                        "Lazy audio loading",
                        isOn: $lazyAudio
                    )

                    Picker(
                        "Data.win loading",
                        selection: $loadType
                    ) {
                        Text("Per chunk")
                            .tag(0)

                        Text("Whole file")
                            .tag(1)
                    }
                }

                // MARK: Display

                Section("Display & Controls") {

                    Toggle(
                        "Touch controls",
                        isOn: $touchControls
                    )

                    Toggle(
                        "FPS",
                        isOn: $showFPS
                    )

                    Toggle(
                        "Widescreen",
                        isOn: $widescreen
                    )

                    if widescreen {
                        Picker(
                            "Aspect ratio",
                            selection: $widescreenAspect
                        ) {
                            Text("16:9")
                                .tag(1.7777778)

                            Text("16:10")
                                .tag(1.6)

                            Text("21:9")
                                .tag(2.3333333)

                            Text("4:3")
                                .tag(1.3333333)
                        }
                    }
                }

                // MARK: Window

                Section("Window Size") {

                    Stepper(
                        "Width: \(windowWidth == 0 ? "Auto" : "\(windowWidth)")",
                        value: $windowWidth,
                        in: 0...7680,
                        step: 1
                    )

                    Stepper(
                        "Height: \(windowHeight == 0 ? "Auto" : "\(windowHeight)")",
                        value: $windowHeight,
                        in: 0...4320,
                        step: 1
                    )

                    if windowWidth != 0 || windowHeight != 0 {
                        Button("Reset to Auto") {
                            windowWidth = 0
                            windowHeight = 0
                        }
                    }
                }

                // MARK: Performance

                Section("Performance") {

                    Picker(
                        "Game speed",
                        selection: $speed
                    ) {
                        Text("0.5×")
                            .tag(0.5)

                        Text("1×")
                            .tag(1.0)

                        Text("1.5×")
                            .tag(1.5)

                        Text("2×")
                            .tag(2.0)

                        Text("3×")
                            .tag(3.0)
                    }
                    .pickerStyle(.segmented)

                    

                }

                Section("Graphics") {
                    VStack(alignment: .leading, spacing: 4) {
                        Toggle(
                            "Metal",
                            isOn: $metal
                        )
                        Text("* Experimental")
                            .font(.caption)
                            .foregroundColor(.secondary)
                        Text("* Graphical glitches may occur.")
                            .font(.caption)
                            .foregroundColor(.secondary)
                        Text("* Better performance")
                            .font(.caption)
                            .foregroundColor(.secondary)

                    }
    

                    VStack(alignment: .leading, spacing: 4) {
                        Toggle(
                            "OpenGL ES",
                            isOn: Binding(
                                get: { !metal },
                                set: { newValue in metal = !newValue }
                            )
                        )
                        Text("* Legacy - will be removed from iOS soon")
                            .font(.caption)
                            .foregroundColor(.secondary)
                        
                        Text("* Better compatibility")
                            .font(.caption)
                            .foregroundColor(.secondary)
                        
                        Text("* Unless using Simulator, no graphical glitches")
                            .font(.caption)
                            .foregroundColor(.secondary)

                        
                    }
                }

                // MARK: Debugging

                Section("Debugging") {

                    Toggle(
                        "Debug mode",
                        isOn: $debug
                    )

                    Toggle(
                        "Disassemble",
                        isOn: $disassemble
                    )

                    Toggle(
                        "Trace frames",
                        isOn: $traceFrames
                    )

                    Toggle(
                        "Trace inherited events",
                        isOn: $traceEventInherited
                    )

                    Toggle(
                        "Log unknown functions",
                        isOn: $alwaysLogUnknownFunctions
                    )

                    Toggle(
                        "Log stubbed functions",
                        isOn: $alwaysLogStubbedFunctions
                    )

                    Toggle(
                        "Disable log colours",
                        isOn: $disableLogColours
                    )
                }

                // MARK: Printing

                Section("Debug Output") {

                    Toggle(
                        "Print rooms",
                        isOn: $printRooms
                    )

                    Toggle(
                        "Print objects",
                        isOn: $printObjects
                    )

                    Toggle(
                        "Print shaders",
                        isOn: $printShaders
                    )

                    Toggle(
                        "Print declared functions",
                        isOn: $printDeclaredFunctions
                    )

                    Toggle(
                        "Print unknown functions",
                        isOn: $printUnknownFunctions
                    )
                }

                // MARK: Frame Control
                
                #if ENABLE_VM_TRACING
                Section("Frame Control") {

                    Stepper(
                        traceBytecodeAfterFrame <= 0
                            ? "Bytecode tracing: Disabled"
                            : "Trace bytecode after frame \(traceBytecodeAfterFrame)",
                        value: $traceBytecodeAfterFrame,
                        in: 0...1_000_000
                    )

                }

                #endif

            }
            .navigationTitle("Settings")
            .toolbar {

                ToolbarItem(
                    placement: .topBarTrailing
                ) {

                    Button("Reset") {
                        reset()
                    }
                }
            }
        }
    }

    private func reset() {

        // Loading
        lazyRooms = true
        lazyTextures = true
        lazyAudio = true
        loadType = 0

        // Display
        touchControls = true
        widescreen = false
        widescreenAspect = 1.7777778
        windowWidth = 0
        windowHeight = 0
        renderer = 0
        headless = false

        // Performance
        speed = 1.0
        fastForwardSpeed = 0.0
        profilerFramesBetween = 0
        opcodeProfiler = false

        // Debugging
        debug = false
        disassemble = false
        alwaysLogUnknownFunctions = false
        alwaysLogStubbedFunctions = false
        disableLogColours = false
        traceFrames = false
        traceEventInherited = false

        // Printing
        printRooms = false
        printObjects = false
        printShaders = false
        printDeclaredFunctions = false
        printUnknownFunctions = false

        // Frame control
        exitAtFrame = -1
        traceBytecodeAfterFrame = 0

        // Seed
        hasSeed = false
        seed = 0

        // Input
        recordInputsPath = ""
        playbackInputsPath = ""

        // Advanced
        eagerRooms = ""
        dumpJsonFilePattern = ""
    }
}

File: android/app/src/main/kotlin/com/butterscotch/butterscotch/SettingsView.kt
package com.butterscotch.butterscotch

import android.content.Context
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Slider
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp

private const val PREFS_NAME = "butterscotch_settings"

@Composable
fun SettingsView(padding: PaddingValues) {
    val context = LocalContext.current

    val prefs = remember {
        context.getSharedPreferences(
            PREFS_NAME,
            Context.MODE_PRIVATE
        )
    }

    // MARK: Loading

    var lazyRooms by remember {
        mutableStateOf(
            prefs.getBoolean("ios.lazyRooms", true)
        )
    }

    var lazyTextures by remember {
        mutableStateOf(
            prefs.getBoolean("ios.lazyTextures", true)
        )
    }

    var lazyAudio by remember {
        mutableStateOf(
            prefs.getBoolean("ios.lazyAudio", true)
        )
    }

    // MARK: Display

    var touchControls by remember {
        mutableStateOf(
            prefs.getBoolean("ios.touchControls", true)
        )
    }

    var showFPS by remember {
        mutableStateOf(
            prefs.getBoolean("ios.showFPS", true)
        )
    }

    var widescreen by remember {
        mutableStateOf(
            prefs.getBoolean("ios.widescreen", false)
        )
    }

    var widescreenAspect by remember {
        mutableFloatStateOf(
            prefs.getFloat(
                "ios.widescreenAspect",
                1.7777778f
            )
        )
    }

    // MARK: Window

    var windowWidth by remember {
        mutableIntStateOf(
            prefs.getInt("ios.windowWidth", 0)
        )
    }

    var windowHeight by remember {
        mutableIntStateOf(
            prefs.getInt("ios.windowHeight", 0)
        )
    }

    // MARK: Performance

    var speed by remember {
        mutableFloatStateOf(
            prefs.getFloat("ios.speed", 1.0f)
        )
    }

    // MARK: Debugging

    var debug by remember {
        mutableStateOf(
            prefs.getBoolean("ios.debug", false)
        )
    }

    var disassemble by remember {
        mutableStateOf(
            prefs.getBoolean("ios.disassemble", false)
        )
    }

    var traceFrames by remember {
        mutableStateOf(
            prefs.getBoolean("ios.traceFrames", false)
        )
    }

    var traceEventInherited by remember {
        mutableStateOf(
            prefs.getBoolean(
                "ios.traceEventInherited",
                false
            )
        )
    }

    var alwaysLogUnknownFunctions by remember {
        mutableStateOf(
            prefs.getBoolean(
                "ios.alwaysLogUnknownFunctions",
                false
            )
        )
    }

    var alwaysLogStubbedFunctions by remember {
        mutableStateOf(
            prefs.getBoolean(
                "ios.alwaysLogStubbedFunctions",
                false
            )
        )
    }

    var disableLogColours by remember {
        mutableStateOf(
            prefs.getBoolean(
                "ios.disableLogColours",
                false
            )
        )
    }

    // MARK: Printing

    var printRooms by remember {
        mutableStateOf(
            prefs.getBoolean("ios.printRooms", false)
        )
    }

    var printObjects by remember {
        mutableStateOf(
            prefs.getBoolean("ios.printObjects", false)
        )
    }

    var printShaders by remember {
        mutableStateOf(
            prefs.getBoolean("ios.printShaders", false)
        )
    }

    var printDeclaredFunctions by remember {
        mutableStateOf(
            prefs.getBoolean(
                "ios.printDeclaredFunctions",
                false
            )
        )
    }

    var printUnknownFunctions by remember {
        mutableStateOf(
            prefs.getBoolean(
                "ios.printUnknownFunctions",
                false
            )
        )
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(padding)
            .verticalScroll(rememberScrollState())
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {

        // MARK: Loading

        SettingsSection("Loading")

        SettingSwitch(
            title = "Lazy room loading",
            checked = lazyRooms,
            onCheckedChange = {
                lazyRooms = it
                prefs.edit()
                    .putBoolean("ios.lazyRooms", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Lazy texture loading",
            checked = lazyTextures,
            onCheckedChange = {
                lazyTextures = it
                prefs.edit()
                    .putBoolean("ios.lazyTextures", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Lazy audio loading",
            checked = lazyAudio,
            onCheckedChange = {
                lazyAudio = it
                prefs.edit()
                    .putBoolean("ios.lazyAudio", it)
                    .apply()
            }
        )

        SettingsDivider()

        // MARK: Display

        SettingsSection("Display & Controls")

        SettingSwitch(
            title = "Touch controls",
            checked = touchControls,
            onCheckedChange = {
                touchControls = it
                prefs.edit()
                    .putBoolean("ios.touchControls", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "FPS",
            checked = showFPS,
            onCheckedChange = {
                showFPS = it
                prefs.edit()
                    .putBoolean("ios.showFPS", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Widescreen",
            checked = widescreen,
            onCheckedChange = {
                widescreen = it
                prefs.edit()
                    .putBoolean("ios.widescreen", it)
                    .apply()
            }
        )

        if (widescreen) {
            SettingChoice(
                title = "Aspect ratio",
                options = listOf(
                    "16:9",
                    "16:10",
                    "21:9",
                    "4:3"
                ),
                selected = when (widescreenAspect) {
                    1.6f -> 1
                    2.3333333f -> 2
                    1.3333333f -> 3
                    else -> 0
                },
                onSelected = {
                    widescreenAspect = when (it) {
                        1 -> 1.6f
                        2 -> 2.3333333f
                        3 -> 1.3333333f
                        else -> 1.7777778f
                    }

                    prefs.edit()
                        .putFloat(
                            "ios.widescreenAspect",
                            widescreenAspect
                        )
                        .apply()
                }
            )
        }

        SettingsDivider()

        // MARK: Window Size

        SettingsSection("Window Size")

        StepperSetting(
            title = "Width",
            value = windowWidth,
            displayValue =
                if (windowWidth == 0) "Auto"
                else "$windowWidth",
            range = 0..7680,
            onValueChange = {
                windowWidth = it

                prefs.edit()
                    .putInt("ios.windowWidth", it)
                    .apply()
            }
        )

        StepperSetting(
            title = "Height",
            value = windowHeight,
            displayValue =
                if (windowHeight == 0) "Auto"
                else "$windowHeight",
            range = 0..4320,
            onValueChange = {
                windowHeight = it

                prefs.edit()
                    .putInt("ios.windowHeight", it)
                    .apply()
            }
        )

        if (windowWidth != 0 || windowHeight != 0) {
            OutlinedButton(
                onClick = {
                    windowWidth = 0
                    windowHeight = 0

                    prefs.edit()
                        .putInt("ios.windowWidth", 0)
                        .putInt("ios.windowHeight", 0)
                        .apply()
                },
                modifier = Modifier.fillMaxWidth()
            ) {
                Text("Reset to Auto")
            }
        }

        SettingsDivider()

        // MARK: Performance

        SettingsSection("Performance")

        Text(
            text = "Game speed: ${speed}×",
            style = MaterialTheme.typography.bodyLarge
        )

        Slider(
            value = speed,
            onValueChange = {
                speed = it

                prefs.edit()
                    .putFloat("ios.speed", it)
                    .apply()
            },
            valueRange = 0.5f..3.0f,
            steps = 4
        )

        SettingsDivider()

        // MARK: Debugging

        SettingsSection("Debugging")

        SettingSwitch(
            title = "Debug mode",
            checked = debug,
            onCheckedChange = {
                debug = it
                prefs.edit()
                    .putBoolean("ios.debug", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Disassemble",
            checked = disassemble,
            onCheckedChange = {
                disassemble = it
                prefs.edit()
                    .putBoolean("ios.disassemble", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Trace frames",
            checked = traceFrames,
            onCheckedChange = {
                traceFrames = it
                prefs.edit()
                    .putBoolean("ios.traceFrames", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Trace inherited events",
            checked = traceEventInherited,
            onCheckedChange = {
                traceEventInherited = it
                prefs.edit()
                    .putBoolean(
                        "ios.traceEventInherited",
                        it
                    )
                    .apply()
            }
        )

        SettingSwitch(
            title = "Log unknown functions",
            checked = alwaysLogUnknownFunctions,
            onCheckedChange = {
                alwaysLogUnknownFunctions = it
                prefs.edit()
                    .putBoolean(
                        "ios.alwaysLogUnknownFunctions",
                        it
                    )
                    .apply()
            }
        )

        SettingSwitch(
            title = "Log stubbed functions",
            checked = alwaysLogStubbedFunctions,
            onCheckedChange = {
                alwaysLogStubbedFunctions = it
                prefs.edit()
                    .putBoolean(
                        "ios.alwaysLogStubbedFunctions",
                        it
                    )
                    .apply()
            }
        )

        SettingSwitch(
            title = "Disable log colours",
            checked = disableLogColours,
            onCheckedChange = {
                disableLogColours = it
                prefs.edit()
                    .putBoolean(
                        "ios.disableLogColours",
                        it
                    )
                    .apply()
            }
        )

        SettingsDivider()

        // MARK: Debug Output

        SettingsSection("Debug Output")

        SettingSwitch(
            title = "Print rooms",
            checked = printRooms,
            onCheckedChange = {
                printRooms = it
                prefs.edit()
                    .putBoolean("ios.printRooms", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Print objects",
            checked = printObjects,
            onCheckedChange = {
                printObjects = it
                prefs.edit()
                    .putBoolean("ios.printObjects", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Print shaders",
            checked = printShaders,
            onCheckedChange = {
                printShaders = it
                prefs.edit()
                    .putBoolean("ios.printShaders", it)
                    .apply()
            }
        )

        SettingSwitch(
            title = "Print declared functions",
            checked = printDeclaredFunctions,
            onCheckedChange = {
                printDeclaredFunctions = it
                prefs.edit()
                    .putBoolean(
                        "ios.printDeclaredFunctions",
                        it
                    )
                    .apply()
            }
        )

        SettingSwitch(
            title = "Print unknown functions",
            checked = printUnknownFunctions,
            onCheckedChange = {
                printUnknownFunctions = it
                prefs.edit()
                    .putBoolean(
                        "ios.printUnknownFunctions",
                        it
                    )
                    .apply()
            }
        )

        SettingsDivider()

        Button(
            onClick = {
                resetSettings(prefs)
            },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("Reset All Settings")
        }
    }
}

@Composable
private fun SettingsSection(title: String) {
    Text(
        text = title,
        style = MaterialTheme.typography.titleLarge,
        modifier = Modifier.padding(
            top = 12.dp,
            bottom = 4.dp
        )
    )
}

@Composable
private fun SettingsDivider() {
    HorizontalDivider(
        modifier = Modifier.padding(
            vertical = 12.dp
        )
    )
}

@Composable
private fun SettingSwitch(
    title: String,
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = title,
            modifier = Modifier.weight(1f)
        )

        Switch(
            checked = checked,
            onCheckedChange = onCheckedChange
        )
    }
}

@Composable
private fun SettingChoice(
    title: String,
    options: List<String>,
    selected: Int,
    onSelected: (Int) -> Unit
) {
    Column {
        Text(
            text = title,
            style = MaterialTheme.typography.bodyLarge
        )

        Row(
            modifier = Modifier.fillMaxWidth()
        ) {
            options.forEachIndexed { index, option ->
                OutlinedButton(
                    onClick = {
                        onSelected(index)
                    },
                    modifier = Modifier.weight(1f)
                ) {
                    Text(option)
                }
            }
        }
    }
}

@Composable
private fun StepperSetting(
    title: String,
    value: Int,
    displayValue: String,
    range: IntRange,
    onValueChange: (Int) -> Unit
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column(
            modifier = Modifier.weight(1f)
        ) {
            Text(title)

            Text(
                text = displayValue,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }

        OutlinedButton(
            onClick = {
                if (value > range.first) {
                    onValueChange(value - 1)
                }
            }
        ) {
            Text("−")
        }

        OutlinedButton(
            onClick = {
                if (value < range.last) {
                    onValueChange(value + 1)
                }
            }
        ) {
            Text("+")
        }
    }
}

private fun resetSettings(
    prefs: android.content.SharedPreferences
) {
    prefs.edit()
        .clear()
        .apply()
}